src/Network.cpp

```diff
@@ -1054,50 +1043,48 @@ void Network::show_heatmap(const FastState* const state,
-            const auto vtx = board.get_vertex(i, j);
-            const auto color = board.get_square(vtx);
-            if (color != FastBoard::EMPTY) {
-                if (color == FastBoard::BLACK) {
-                    black[idx] = true;
-                } else {
-                    white[idx] = true;
-                }
-            }
-            idx++;
-        }
-    }
-}
-
-void Network::gather_features(const GameState* const state, NNPlanes & planes) {
-    planes.resize(INPUT_CHANNELS);
-    auto& black_to_move = planes[2 * INPUT_MOVES];
-    auto& white_to_move = planes[2 * INPUT_MOVES + 1];
-
-    const auto to_move = state->get_to_move();
-    const auto blacks_move = to_move == FastBoard::BLACK;
+void Network::gather_features(const GameState* const state,
+                              std::vector<net_t>& input_data,
+                              const int symmetry) {
+    assert(symmetry >= 0 && symmetry <= 7);
 
-    const auto black_offset = blacks_move ? 0 : INPUT_MOVES;
-    const auto white_offset = blacks_move ? INPUT_MOVES : 0;
+    const auto color_to_move = state->get_to_move();
+    const auto color_not_to_move = color_to_move == FastBoard::BLACK ?
+                                                    FastBoard::WHITE :
+                                                    FastBoard::BLACK;
+    auto color_not_to_move_data = std::vector<net_t>{};
 
-    if (blacks_move) {
-        black_to_move.set();
-    } else {
-        white_to_move.set();
-    }
+    input_data.reserve(INPUT_CHANNELS * BOARD_SQUARES);
+    color_not_to_move_data.reserve(INPUT_MOVES * BOARD_SQUARES);
 
     const auto moves = std::min<size_t>(state->get_movenum() + 1, INPUT_MOVES);
-    // Go back in time, fill history boards
-    for (auto h = size_t{0}; h < moves; h++) {
-        // collect white, black occupation planes
-        fill_input_plane_pair(state->get_past_board(h),
-                              planes[black_offset + h],
-                              planes[white_offset + h]);
+    for (auto c = size_t{0}; c < moves; c++) {
+        auto const board = state->get_past_board(c);
+        for (auto idx = 0; idx < BOARD_SQUARES; idx++) {
+            const auto sym_idx = symmetry_nn_idx_table[symmetry][idx];
+            const auto x = sym_idx % BOARD_SIZE;
+            const auto y = sym_idx / BOARD_SIZE;
+            const auto color = board.get_square(x, y);
+            input_data.emplace_back(
+                net_t(color == color_to_move));
+            color_not_to_move_data.emplace_back(
+                net_t(color == color_not_to_move));
+        }
+    }
+    for (auto c = moves; c < INPUT_MOVES; c++) {
+        input_data.insert(cend(input_data), BOARD_SQUARES, net_t(false));
+        color_not_to_move_data.insert(cend(color_not_to_move_data),
+            BOARD_SQUARES, net_t(false));
     }
+
+    input_data.insert(cend(input_data),
+        std::make_move_iterator(cbegin(color_not_to_move_data)),
```

Comment:
Given that the planes are single-dimensional floating point vectors, I don't see how you're actually going to have any possibility of moving here. You end up copying all the data from your temporary array back to the original one.

To avoid this, I think it's better to calculate the right offset for color_not_to_move, resize input_data suitably in the beginning, and use the offsets for directly storing instead of emplace_back. 

Your previous version avoided the copy so I think this now too clever.

Reply:
Yeah, I'm still learning :) I'll put it back the way it used to be done. Thanks for the useful info again..